File: backend-fastapi/app/api/routes/license.py
"""
Auto-converted from Flask Blueprint → FastAPI APIRouter
Source files: license.py
"""
import logging
from typing import Optional
from fastapi import APIRouter, HTTPException, Query, Request, Depends, UploadFile, File
from fastapi.responses import JSONResponse, StreamingResponse
from core.config import settings
from core.dependencies import get_current_user, get_admin_user

logger = logging.getLogger(__name__)

router = APIRouter()

# ============================================================
# Source: license.py
# ============================================================

"""
许可证API路由

处理与系统许可证相关的所有API请求
"""
import json
from app.services.license_service import LicenseService

# 创建Blueprint

@router.get('/license')
def get_license():
    """获取当前许可证信息"""
    # 创建服务实例
    license_service = LicenseService()
    license_info = license_service.get_current_license()
    if not license_info:
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '未找到有效的许可证',
            'code': 'LICENSE_EXPIRED'
        })

    return {
        'status': 'success',
        'data': license_info
    }

@router.get('/license/expired')
def get_expired_license():
    """获取过期的许可证信息（即使已过期也返回）"""
    # 创建服务实例
    license_service = LicenseService()
    license_info = license_service.get_license_data(include_expired=True)

    if not license_info:
        raise HTTPException(status_code=404, detail={
            'status': 'error',
            'message': '未找到任何许可证信息'
        })

    return {
        'status': 'success',
        'data': license_info
    }

@router.post('/license/activate')
async def activate_license(request: Request):
    """通过密钥激活许可证"""
    data = await request.json()
    license_key = data.get('license_key')

    if not license_key:
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '缺少许可证密钥'
        })

    # 创建服务实例
    license_service = LicenseService()
    result = license_service.activate_license(license_key)

    if not result['success']:
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': result['message']
        })

    return {
        'status': 'success',
        'data': result['license']
    }

@router.post('/license/activate-file')
async def activate_license_file(license_file: UploadFile = File(...)):
    """通过文件激活许可证"""
    # 检查文件名
    if not license_file.filename or license_file.filename == '':
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '未选择文件'
        })

    # 检查文件类型
    if not license_file.filename.endswith('.json'):
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '许可证文件必须是JSON格式'
        })

    try:
        # 读取文件内容
        license_data = json.loads((await license_file.read()).decode('utf-8'))

        # 检查文件格式
        if 'license_key' not in license_data:
            raise HTTPException(status_code=400, detail={
                'status': 'error',
                'message': '无效的许可证文件格式'
            })

        # 创建服务实例并激活许可证
        license_service = LicenseService()
        result = license_service.activate_license(license_data['license_key'])

        if not result['success']:
            raise HTTPException(status_code=400, detail={
                'status': 'error',
                'message': result['message']
            })

        return {
            'status': 'success',
            'data': result['license']
        }
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '无效的JSON文件'
        })
    except Exception as e:
        logger.error(f"通过文件激活许可证失败: {e}")
        return JSONResponse(content={
            'status': 'error',
            'message': f'激活许可证失败: {str(e)}'
        }, status_code=500)

@router.get('/license/check-feature')
def check_feature(request: Request):
    """检查功能是否可用"""
    feature_name = request.query_params.get('feature')

    if not feature_name:
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '缺少功能名称'
        })

    # 创建服务实例
    license_service = LicenseService()
    available = license_service.check_feature_availability(feature_name)

    return {
        'status': 'success',
        'data': {
            'feature': feature_name,
            'available': available
        }
    }

@router.get('/license/check-limit')
def check_limit(request: Request):
    """检查资源限制"""
    resource_type = request.query_params.get('resource')
    current_count = int(request.query_params.get('count', '0'))

    if not resource_type:
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '缺少资源类型'
        })

    if resource_type not in ['agents', 'action_spaces', 'roles']:
        return JSONResponse(content={
            'status': 'error',
            'message': f'不支持的资源类型: {resource_type}'
        }, status_code=400)

    # 创建服务实例
    license_service = LicenseService()
    allowed = license_service.check_resource_limit(resource_type, current_count)

    return {
        'status': 'success',
        'data': {
            'resource': resource_type,
            'current_count': current_count,
            'allowed': allowed
        }
    }

@router.get('/license/system-key')
def get_system_key():
    """获取系统许可证密钥"""
    # 从系统设置中获取密钥
    from app.models import SystemSetting
    secret_key = SystemSetting.get('license_secret_key')

    if not secret_key:
        # 如果没有找到密钥，返回错误
        raise HTTPException(status_code=400, detail={
            'status': 'error',
            'message': '系统未配置许可证密钥'
        })

    # 返回密钥
    return {
        'status': 'success',
        'data': {
            'key': secret_key
        }
    }

File: frontend/src/services/api/license.ts
import api from './axios';

/**
 * 许可证API服务
 */
export const licenseAPI = {
  /**
   * 获取当前许可证信息
   * @returns {Promise<Object>} 许可证信息
   */
  getCurrentLicense: async () => {
    try {
      // 确保使用正确的路径，不要在末尾添加斜杠
      console.log('正在请求许可证信息，URL: /license');
      const response = await api.get('/license');
      console.log('许可证信息请求成功:', response.config.url);
      return response.data.data;
    } catch (error) {
      console.error('获取许可证信息失败:', error);
      console.error('请求URL:', error.config?.url);
      throw error;
    }
  },

  /**
   * 获取过期的许可证信息
   * @returns {Promise<Object>} 过期的许可证信息
   */
  getExpiredLicense: async () => {
    try {
      const response = await api.get('/license/expired');
      return response.data.data;
    } catch (error) {
      console.error('获取过期许可证信息失败:', error);
      // 对于404错误，抛出特定的license错误
      if (error.response?.status === 404) {
        const licenseError: any = new Error('未找到任何许可证信息');
        licenseError.isLicenseError = true;
        licenseError.code = 'LICENSE_NOT_FOUND';
        throw licenseError;
      }
      throw error;
    }
  },

  /**
   * 通过密钥激活许可证
   * @param {string} licenseKey 许可证密钥
   * @returns {Promise<Object>} 激活结果
   */
  activateLicense: async (licenseKey) => {
    try {
      const response = await api.post('/license/activate', { license_key: licenseKey });
      return response.data;
    } catch (error) {
      console.error('激活许可证失败:', error);
      throw error;
    }
  },

  /**
   * 通过文件激活许可证
   * @param {File} file 许可证文件
   * @returns {Promise<Object>} 激活结果
   */
  activateLicenseFile: async (file) => {
    try {
      const formData = new FormData();
      formData.append('license_file', file);

      const response = await api.post('/license/activate-file', formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      });

      return response.data;
    } catch (error) {
      console.error('通过文件激活许可证失败:', error);
      throw error;
    }
  },

  /**
   * 检查功能是否可用
   * @param {string} featureName 功能名称
   * @returns {Promise<boolean>} 功能是否可用
   */
  checkFeatureAvailability: async (featureName) => {
    try {
      const response = await api.get(`/license/check-feature?feature=${featureName}`);
      return response.data.data.available;
    } catch (error) {
      console.error(`检查功能 ${featureName} 可用性失败:`, error);
      return false;
    }
  },

  /**
   * 检查资源限制
   * @param {string} resourceType 资源类型 (agents, action_spaces, roles)
   * @param {number} currentCount 当前资源数量
   * @returns {Promise<boolean>} 是否允许创建更多资源
   */
  checkResourceLimit: async (resourceType, currentCount) => {
    try {
      const response = await api.get(`/license/check-limit?resource=${resourceType}&count=${currentCount}`);
      return response.data.data.allowed;
    } catch (error) {
      console.error(`检查资源 ${resourceType} 限制失败:`, error);
      return false;
    }
  },

  /**
   * 获取系统许可证密钥
   * @returns {Promise<string>} 系统许可证密钥
   */
  getSystemKey: async () => {
    try {
      const response = await api.get('/license/system-key');
      return response.data.data.key;
    } catch (error) {
      console.error('获取系统许可证密钥失败:', error);
      throw error; // 不提供默认密钥，直接抛出错误
    }
  }
};

export default licenseAPI;
